### PR TITLE
panic_str only exists for the migration to 2021 panic macros

### PR DESCRIPTION
**Pull request by RalfJung**
The only caller is `expect_failed`, which is already a cold inline(never) function, so inlining into that function should be fine. (And indeed `panic_str` was `#[inline]` anyway.)

The existence of panic_str risks someone calling it when they should call `panic` instead, and I can't see a reason why this footgun should exist.

I also extended the comment in `panic` to explain why it needs a `'static` string -- I know I've wondered about this in the past and it took me quite a while to understand.

### CODE REVIEW

**Comment by m-ou-se** [comment]
`panic_str` is also used here: https://github.com/rust-lang/rust/blob/60b5ca62752ecc25d578066c8b82e1a4887267d4/library/core/src/panic.rs#L30-L33

**Comment by RalfJung** [comment]
Ah, "find references" led me astray...

I originally just wanted to add some comments, I will probably turn the PR into that then. (Assuming that lint can't easily access type information.)

**Comment by RalfJung** [comment]
It seems like that lint does have type information available via `cx.typeck_results().expr_ty`. So instead of checking for `panic_str` it could check for `panic_display` with an `&&str` argument -- or would that not give the right results?

**Comment by RalfJung** [comment]
> It seems like that lint does have type information available via `cx.typeck_results().expr_ty`. So instead of checking for `panic_str` it could check for `panic_display` with an `&&str` argument -- or would that not give the right results?

I guess that would not be right since the lint should kick in for `println!(mystrvar)` but not `println!("{}", mystrvar)`.

File: tests/ui-fulldeps/stable-mir/check_transform.rs [hunk 1]
```diff
@@ -137,9 +137,9 @@ fn generate_input(path: &str) -> std::io::Result<()> {
     write!(
         file,
         r#"
-        #![feature(panic_internals)]
```
Comment (RalfJung) [inline tests/ui-fulldeps/stable-mir/check_transform.rs:140]: Was it really necessary to make this test use extremely unstable *internal* features? I hope we can avoid that, it just makes unnecessary work for others when those internal details get changed.

Comment (RalfJung) [reply tests/ui-fulldeps/stable-mir/check_transform.rs:140]: Cc @celinval

Comment (celinval) [reply tests/ui-fulldeps/stable-mir/check_transform.rs:140]: Fair.. sorry

**Review by m-ou-se — APPROVED** [review]
Lgtm! r=me when ready.

**Comment by RalfJung** [comment]
This is ready, I just rebased to fix a typo in the commit message. :)

@bors r=m-ou-se  rollup

**Pull request merged**